File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys

sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline


def dfs(graph, x, y, n):
    graph[x][y] = -1  # 방문 표시
    global count
    count += 1
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx < n and 0 <= ny < n and (graph[nx][ny] == 1):
            dfs(graph, nx, ny, n)


n = int(input())
graph = [list(map(int, input().strip())) for _ in range(n)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
count = 0

home_count = []
for i in range(n):
    for j in range(n):
        # 돌면서 안 돈 곳 있으면 돌아야 함
        if graph[i][j] == 1:
            dfs(graph, i, j, n)
            home_count.append(count)
            count = 0
home_count.sort()
print(len(home_count))
print('\n'.join(map(str, home_count)))

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.kt
package org.example

import java.util.StringTokenizer

private val dx = intArrayOf(-1, 1, 0, 0)
private val dy = intArrayOf(0, 0, -1, 1)

fun main() {
    val br = System.`in`.bufferedReader()
    val N = br.readLine().toInt()
    val graph = Array(N) {
        br.readLine().map { it.digitToInt() }.toIntArray()
    }

    // 단지 번호 붙여서 단지 수
    val homeCountList = mutableListOf<Int>()
    for (i in graph.indices) {
        for (j in graph[0].indices) {
            if (graph[i][j] == 1) {
                homeCountList.add(dfs(graph, i, j))
            }
        }
    }

    homeCountList.sort()

    val sb = StringBuilder()
    sb.append(homeCountList.size).append("\n")
    homeCountList.forEach { sb.append(it).append("\n") }
    print(sb)

    br.close()
}

fun dfs(graph: Array<IntArray>, x: Int, y: Int): Int {
    var count = 1
    graph[x][y] = 0

    for (i in 0 until 4) {
        val nx = x + dx[i]
        val ny = y + dy[i]
        if (nx in graph.indices && ny in graph[0].indices && graph[nx][ny] == 1) {
            count += dfs(graph, nx, ny)
        }
    }

    return count
}